Code/Tools/AssetProcessor/native/utilities/assetUtils.cpp

```diff
@@ -1426,22 +1427,37 @@ namespace AssetUtilities
         return source;
     }
 
-    AZStd::vector<AZStd::string> GetAllIntermediateSources(
-        AZ::IO::PathView relativeSourcePath, AZStd::shared_ptr<AssetProcessor::AssetDatabaseConnection> db)
+    AZStd::vector<AssetProcessor::SourceAssetReference> GetAllIntermediateSources(
+        const AssetProcessor::SourceAssetReference& sourceAsset, AZStd::shared_ptr<AssetProcessor::AssetDatabaseConnection> db)
     {
-        AZStd::vector<AZStd::string> sources;
+        AZStd::vector<AssetProcessor::SourceAssetReference> sources;
 
-        auto topLevelSource = GetTopLevelSourceForProduct(relativeSourcePath, db);
+        auto topLevelSource = GetTopLevelSourceForIntermediateAsset(sourceAsset, db);
 
         if (!topLevelSource)
         {
-            AzToolsFramework::AssetDatabase::SourceDatabaseEntry source;
-            db->GetSourceBySourceName(relativeSourcePath.FixedMaxPathStringAsPosix().c_str(), source);
+            AzToolsFramework::AssetDatabase::SourceDatabaseEntryContainer source;
+            db->GetSourcesBySourceNameScanFolderId(sourceAsset.RelativePath().c_str(), sourceAsset.ScanfolderId(), source);
 
-            topLevelSource = source;
+            AZ_Assert(
```

Comment:
This is data coming in from the database and it can AZ_Assert
If this happens, will the user just need to nuke the database because the code will just assert each time now.

Reply:
This API is honestly poorly designed.  It makes no sense for it to return a container of sources when it should be impossible to have more than 1 source with the same name and scanfolder ID.  That would mean 2 files exist at the same absolute path, which could only happen if the database is corrupted or some code screwed up.

I could remove the assert I suppose, since I doubt it will ever fire, I think I just felt it explained the assumption made below where I'm just blindly grabbing index 0.

Reply:
So here. GetSourcesBySourceNameScanFolderId() comes back with a SourceDatabaseEntryContainer named source. This is incoming data from the data base. If the "source.size() > 1" then the assert will fire and kill the AP due to this assert. 

This seems to be triggering an assert based on input data which means what for the user? How will coding fix this?

Reply:
I've removed the assert since it is an impossible situation and having more than 1 result won't affect anything